sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -158,48 +175,63 @@ private static void WriteMemory(MemoryStream stream, ReadOnlyMemory<byte> memory
         /// ]]></code>
         /// </example>
         /// <remarks> See error response code and error response message for more detail. </remarks>
-        public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string streamName, IEnumerable<T> logEntries, CancellationToken cancellationToken = default)
+        public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string streamName, IEnumerable<T> logEntries, UploadLogsOptions options = null, CancellationToken cancellationToken = default)
         {
             Argument.AssertNotNullOrEmpty(ruleId, nameof(ruleId));
             Argument.AssertNotNullOrEmpty(streamName, nameof(streamName));
             Argument.AssertNotNullOrEmpty(logEntries, nameof(logEntries));
 
             using var scope = ClientDiagnostics.CreateScope("LogsIngestionClient.Upload");
-            scope.Start();
 
             RequestContext requestContext = GenerateRequestContext(cancellationToken);
             Response response = null;
             List<UploadLogsError> errors = new List<UploadLogsError>();
+            //TODO: use options.Serializer
             try
             {
+                scope.Start();
+                // A list of tasks that are currently executing which will
+                // always be smaller than MaxWorkerCount
+                List<Task<RunningTask<T>>> runningTasks = new();
+                // Partition the stream into individual blocks
                 foreach (BatchedLogs<T> batch in Batch(logEntries))
                 {
-                    //TODO: catch errors and correlate with Batch.start
-                    using HttpMessage message = CreateUploadRequest(ruleId, streamName, batch.LogsData, "gzip", requestContext);
-                    response = _pipeline.ProcessMessage(message, requestContext, cancellationToken);
-                    if (response.Status != 204) // if any error is thrown log it
+                    // Because we are uploading in sequence, wait for each batch to upload before starting the next batch
```

Comment:
Why are we uploading in sequence? Does the service expects the order to be always maintained?

Reply:
the non-async method will run in sequence - ensure one batch has been uploaded before moving to the next. Sync methods will have just the one thread